src/CSnakes.Runtime/PyObjectTypeConverter.Tuple.cs

```diff
@@ -1,81 +1,75 @@
-        {
-            PyObject value = PyObject.Create(CPythonAPI.PyTuple_GetItem(pyObj, i));
-            tupleValues.Add(value);
-        }
-
-        // If we have more than 8 python values, we are going to have a nested tuple, which we have to unpack.
-        if (tupleValues.Count > 8)
-        {
-            // We are hitting nested tuples here, which will be treated in a different way.
-            object?[] firstSeven = tupleValues.Take(7).Select((p, i) => ConvertTo(context, culture, p, types[i])).ToArray();
-
-            // Get the rest of the values and convert them to a nested tuple.
-            IEnumerable<PyObject> rest = tupleValues.Skip(7);
-
-            // Back to a Python tuple.
-            using PyObject pyTuple = PyTuple.CreateTuple(rest);
-
-            // Use the decoder pipeline to decode the nested tuple (and its values).
-            // We do this because that means if we have nested nested tuples, they'll be decoded as well.
-            object? nestedTuple = ConvertTo(context, culture, pyTuple, types[7]);
-
-            // Append our nested tuple to the first seven values.
-            clrValues = [.. firstSeven, nestedTuple];
-        }
-        else
-        {
-            clrValues = tupleValues.Select((p, i) => ConvertTo(context, culture, p, types[i])).ToArray();
-        }
-
-        // Dispose of all the Python values that we captured from the Tuple.
-        foreach (var value in tupleValues)
-        {
-            value.Dispose();
+using CSnakes.Runtime.CPython;
+using CSnakes.Runtime.Python;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+namespace CSnakes.Runtime;
+internal partial class PyObjectTypeConverter
+{
+    private PyObject ConvertFromTuple(ITuple t)
+    {
+        List<PyObject> pyObjects = new(t.Length);
+
+        for (var i = 0; i < t.Length; i++)
+        {
+            pyObjects.Add(ToPython(t[i]));
+        }
+
+        return Pack.CreateTuple(pyObjects);
+    }
+
+    private object? ConvertToTuple(PyObject pyObj, Type destinationType)
+    {
+        // We have to convert the Python values to CLR values, as if we just tried As<object>() it would
+        // not parse the Python type to a CLR type, only to a new Python type.
+        Type[] types = destinationType.GetGenericArguments();
+        nint tupleSize = CPythonAPI.PyTuple_Size(pyObj);
+        object?[] clrValues = new object[Math.Min(8, tupleSize)];
+
+        PyObject[] tupleValues = new PyObject[tupleSize];
+        for (nint i = 0; i < tupleValues.Length; i++)
+        {
+            PyObject value = PyObject.Create(CPythonAPI.PyTuple_GetItemWithNewRef(pyObj, i));
+            tupleValues[i] = value;
+        }
+
+        // If we have more than 8 python values, we are going to have a nested tuple, which we have to unpack.
+        if (tupleValues.Length > 8)
+        {
+            // We are hitting nested tuples here, which will be treated in a different way.

```

Comment:
Unclear if this is a hot path, but this style of LINQ is a bit more expensive than a simple for loop and passing around a sliced `Span<object>`.